asm_utils/src/ast.rs

```diff
@@ -0,0 +1,200 @@
+//! Common AST for the frontend architecture inputs.
+
+use std::fmt::{self, Display};
+
+#[derive(Clone)]
+pub enum Statement<R: Register, F: FunctionOpKind> {
+    Label(String),
+    Directive(String, Vec<Argument<R, F>>),
+    Instruction(String, Vec<Argument<R, F>>),
+}
+
+#[derive(Clone)]
+pub enum Argument<R: Register, F: FunctionOpKind> {
+    Register(R),
+    RegOffset(R, Expression<F>),
+    StringLiteral(Vec<u8>),
+    Expression(Expression<F>),
+}
+
+impl<R: Register, F: FunctionOpKind> Argument<R, F> {
+    pub fn post_visit_expressions_mut(&mut self, f: &mut impl FnMut(&mut Expression<F>)) {
+        match self {
+            Argument::Register(_) | Argument::StringLiteral(_) => (),
+            Argument::RegOffset(_, expr) | Argument::Expression(expr) => {
+                expr.post_visit_mut(f);
+            }
+        }
+    }
+
+    pub fn post_visit_expressions<'a>(&'a self, f: &mut impl FnMut(&'a Expression<F>)) {
+        match self {
+            Argument::Register(_) | Argument::StringLiteral(_) => (),
+            Argument::RegOffset(_, expr) | Argument::Expression(expr) => {
+                expr.post_visit(f);
+            }
+        }
+    }
+}
+
+pub trait Register: Display {}
+
+#[derive(Clone, Copy)]
+pub enum UnaryOpKind {
+    Negation,
+}
+
+#[derive(Clone, Copy)]
+pub enum BinaryOpKind {
+    Or,
+    Xor,
+    And,
+    LeftShift,
+    RightShift,
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Mod,
+}
+
+pub trait FunctionOpKind: Display {}
+
+#[derive(Clone)]
+pub enum Expression<F> {
+    Number(i64),
+    Symbol(String),
+    UnaryOp(UnaryOpKind, Box<[Expression<F>]>),
```

Comment:
I think we should make the arguments just single expressions (both for UnaryOp as well as for FunctionOp) instead of arrays. We can change it again when we need more arguments, but currently I think it's more dangerous to allow arrays but then assume that they have size one.